=== modified file 'Makefile'
--- upstream/Makefile	2009-12-21 19:01:26 +0000
+++ regex/Makefile	2009-12-21 19:15:12 +0000
@@ -33,6 +33,7 @@
 DNSMASQ_CFLAGS=`echo $(COPTS) | ../bld/pkg-wrapper HAVE_DBUS $(PKG_CONFIG) --cflags dbus-1` 
 DNSMASQ_LIBS=  `echo $(COPTS) | ../bld/pkg-wrapper HAVE_DBUS $(PKG_CONFIG) --libs dbus-1` 
 SUNOS_LIBS= `if uname | grep SunOS 2>&1 >/dev/null; then echo -lsocket -lnsl -lposix4; fi`
+DNSMASQ_LIBS +=`$(PKG_CONFIG) --libs libpcre`
 
 all :   dnsmasq
 

=== modified file 'src/dnsmasq.h'
--- upstream/src/dnsmasq.h	2009-12-21 19:06:28 +0000
+++ regex/src/dnsmasq.h	2009-12-21 19:09:21 +0000
@@ -110,6 +110,8 @@
 #include <priv.h>
 #endif
 
+#include <pcre.h>
+
 /* daemon is function in the C library.... */
 #define daemon dnsmasq_daemon
 
@@ -309,6 +311,7 @@
 #define SERV_MARK            256  /* for mark-and-delete */
 #define SERV_TYPE    (SERV_HAS_DOMAIN | SERV_FOR_NODOTS)
 #define SERV_COUNTED         512  /* workspace for log code */
+#define SERV_IS_REGEX       1024  /* server entry is a regex */
 
 struct serverfd {
   int fd;
@@ -327,6 +330,8 @@
   char interface[IF_NAMESIZE+1];
   struct serverfd *sfd; 
   char *domain; /* set if this server only handles a domain. */ 
+  pcre *regex;
+  pcre_extra *pextra;
   int flags, tcpfd;
   unsigned int queries, failed_queries;
   struct server *next; 

=== modified file 'src/forward.c'
--- upstream/src/forward.c	2009-12-21 19:01:26 +0000
+++ regex/src/forward.c	2009-12-21 19:09:21 +0000
@@ -149,12 +149,36 @@
       }
     else if (serv->flags & SERV_HAS_DOMAIN)
       {
-	unsigned int domainlen = strlen(serv->domain);
-	char *matchstart = qdomain + namelen - domainlen;
-	if (namelen >= domainlen &&
-	    hostname_isequal(matchstart, serv->domain) &&
-	    domainlen >= matchlen &&
-	    (domainlen == 0 || namelen == domainlen || *(serv->domain) == '.' || *(matchstart-1) == '.' ))
+	unsigned int domainlen = matchlen;
+	int serverhit = 0;
+
+	if (!(serv->flags & SERV_IS_REGEX))
+	  {
+	    domainlen = strlen(serv->domain);
+	    char *matchstart = qdomain + namelen - domainlen;
+	    if (namelen >= domainlen &&
+	        hostname_isequal(matchstart, serv->domain) &&
+	        domainlen >= matchlen &&
+	        (domainlen == 0 || namelen == domainlen || *(serv->domain) == '.' || *(matchstart-1) == '.' ))
+	       serverhit = 1;
+	  }
+	else
+	  {
+	    int captcount = 0;
+	    if (pcre_fullinfo(serv->regex, serv->pextra, PCRE_INFO_CAPTURECOUNT, &captcount) == 0)
+	      {
+		/* C99 dyn-array, or alloca must be used */
+		int ovect[(captcount + 1) * 3];
+		if (pcre_exec(serv->regex, serv->pextra, qdomain, namelen, 0, 0, ovect, (captcount + 1) * 3) > 0)
+		  {
+		    domainlen = (unsigned int) (ovect[1] - ovect[0]);
+		    if (domainlen >= matchlen)
+		      serverhit = 1;
+		  }
+	      }
+	  }
+
+	if (serverhit)
 	  {
 	    unsigned short sflag = serv->addr.sa.sa_family == AF_INET ? F_IPV4 : F_IPV6;
 	    *type = SERV_HAS_DOMAIN;

=== modified file 'src/network.c'
--- upstream/src/network.c	2009-12-21 19:01:26 +0000
+++ regex/src/network.c	2009-12-21 19:09:21 +0000
@@ -718,7 +718,7 @@
 	  char *s1, *s2;
 	  if (!(new->flags & SERV_HAS_DOMAIN))
 	    s1 = _("unqualified"), s2 = _("names");
-	  else if (strlen(new->domain) == 0)
+	  else if (new->domain && strlen(new->domain) == 0)
 	    s1 = _("default"), s2 = "";
 	  else
 	    s1 = _("domain"), s2 = new->domain;

=== modified file 'src/option.c'
--- upstream/src/option.c	2009-12-21 19:01:26 +0000
+++ regex/src/option.c	2009-12-21 19:22:18 +0000
@@ -1391,10 +1391,16 @@
 	    arg++;
 	    while ((end = split_chr(arg, '/')))
 	      {
-		char *domain = NULL;
+		char *domain = NULL, *regex = NULL;
+		char *real_end = arg + strlen(arg);
 		/* # matches everything and becomes a zero length domain string */
 		if (strcmp(arg, "#") == 0)
 		  domain = "";
+		else if (*arg == ':' && *(real_end - 1) == ':')
+		  {
+		     *(real_end - 1) = '\0';
+		     regex = arg + 1;
+		  }
 		else if (strlen (arg) != 0 && !(domain = canonicalise_opt(arg)))
 		  option = '?';
 		serv = opt_malloc(sizeof(struct server));
@@ -1402,7 +1408,22 @@
 		serv->next = newlist;
 		newlist = serv;
 		serv->domain = domain;
-		serv->flags = domain ? SERV_HAS_DOMAIN : SERV_FOR_NODOTS;
+		serv->flags = domain || regex ? SERV_HAS_DOMAIN : SERV_FOR_NODOTS;
+		if (regex)
+		  {
+		    const char *error;
+		    int erroff;
+		    serv->regex = pcre_compile(regex, 0, &error, &erroff, NULL);
+
+		    if (!serv->regex)
+		      {
+			 option = '?';
+			 problem = (char *) error;
+			 break;
+		      }
+		    serv->flags |= SERV_IS_REGEX;
+		    serv->pextra = pcre_study(serv->regex, 0, &error);
+		  }
 		arg = end;
 	      }
 	    if (!newlist)

